Implement bitvectors

mpardalos opened the pull request:
Adding in support for bitvectors, as discussed in https://github.com/smtcoq/smtcoq/issues/150.

mpardalos committed:
Implement bitvectors

ckeller commented:
Thanks @mpardalos for starting this work! Do not hesitate to tell me if you have any question or request.

mpardalos committed:
Add nat2bv function (probably broken)

mpardalos commented:
Could you give me some instructions for how to build the ocaml part of the library? This is what I get if I just run make in the root folder:

```
❯ make 
make: TMPDIR value /tmp/nix-shell-189000-0: No such file or directory
make: using default temporary directory '/tmp'
COQPP src/g_api.mlg
CAMLDEP src/g_api.ml
CAMLDEP src/api.ml
OCAMLLIBDEP src/smtcoq_api_plugin.mlpack
CAMLDEP src/api.mli
COQDEP VFILES
make: TMPDIR value /tmp/nix-shell-189000-0: No such file or directory
make: using default temporary directory '/tmp'
make: TMPDIR value /tmp/nix-shell-189000-0: No such file or directory
make: using default temporary directory '/tmp'
make: TMPDIR value /tmp/nix-shell-189000-0: No such file or directory
make: using default temporary directory '/tmp'
COQC theories/SMTLib.v
CAMLC -c src/api.mli
ocamlfind: /tmp/nix-shell-189000-0/findlib_initl34205a.ml: No such file or directory
make[1]: *** [Makefile.coq:751: src/api.cmi] Error 2
make: *** [Makefile.coq:417: all] Error 2
```

I have installed the coq/ocaml tools with nix, using this shell.nix file. Maybe there is some dependencies missing?
```nix
{ nixpkgs ? import (fetchTarball
  "https://github.com/NixOS/nixpkgs/archive/fecfeb86328381268e29e998ddd3ebc70bbd7f7c.tar.gz")
}:
let
  pkgs = nixpkgs {};
  coq = pkgs.coq_8_19;
  coqPackages = pkgs.coqPackages_8_19;
in pkgs.mkShell {
  packages = [
    coq
    coqPackages.smtcoq

    coq.ocaml
    coq.ocamlPackages.findlib
    coq.ocamlPackages.ocaml-lsp
    coq.ocamlPackages.dune_3
    coq.ocamlPackages.utop
    coq.ocamlPackages.ocamlformat
    coq.ocamlPackages.menhir
    coq.ocamlPackages.ppx_deriving
    coq.ocamlPackages.ppxlib
    coq.ocamlPackages.yojson
  ];
}

```

ckeller commented:
I am sorry but I am not familiar at all with Nix.
It looks like `make` would like to use `/tmp/nix-shell-189000-0` as a temporary directory but it does not exist, so it uses `/tmp`, but still it searches for `findlib`in `/tmp/nix-shell-189000-0`.
Maybe ask on a nix forum?

mpardalos committed:
Add some relevant boolean functions

Add Term_Not
Add Term_ITE
Add Term_Or
Add Term_True and Term_False

ckeller commented:
@mpardalos Did you manage to solve the nix problem?

mpardalos commented:
Yes, I managed to solve it, apparently I just needed to set TMPDIR manually. I can build the project just fine now.

I've been adding things to the term type as I need them for my project. What do you think would be needed to merge these changes back into master?

ckeller commented:
To be able to generate SMT-LIB files out of your work, the OCaml part should also be extended.
It means exporting the new sorts and terms you have defined as done [here](https://github.com/mpardalos/smtcoq-api/blob/bitvectors/theories/SMTLib.v#L393), and handling them in OCaml in [this file](https://github.com/mpardalos/smtcoq-api/blob/bitvectors/src/api.ml). Do you need help for that?
Then the extension can be tested as presented in [the example](https://github.com/mpardalos/smtcoq-api/blob/bitvectors/examples/Example.v).

ckeller commented:
Thanks for the commit!
Would you like to export queries rather than terms directly?

mpardalos committed:
WIP implementing ocaml API

mpardalos commented:
I have been pushing some more changes to my fork, which are relevant for my work (including the query type), but I think it's best if we merge the bitvector changes since those are most likely to be useful for other people too.

I have kept all bitvector-related changes on this branch and moved everything else to `extras`. 

I am having some trouble implementing the OCaml side of this, because I don't have any experience with Coq plugins or SMTCoq internals. I think for most of what I've added I can just follow the pattern from the code that is already there, but I'm not sure how to implement `compile_sort` for the `Sort_BitVec` that I've added. I have left a relevant comment in the source. Could I get some pointers for how to proceed with that?

ckeller commented:
Dear @mpardalos 
Thank you very much for all your work.
I will come back in a few days with more details about the OCaml part after a careful reading of what you have already done.

ckeller commented:
Hi @mpardalos 
Sorry for answering so late but it was a heavy week for me.
In the OCaml code, `Sort_Uninterpreted`takes a `Constr.t` (that is to say a Rocq term, in its OCaml representation) as a parameter because it represents some Rocq sort that the SMT solver does not know about. On the contrary, `Sort_BitVec` is interpreted si it can take an OCaml `ìnt` as a parameter, that can be manipulated easily in OCaml but also translated into a Coq term if needed.
Does that make sense?

ckeller committed:
Adding BVLit in OCaml, with an example

ckeller commented:
I have proposed a PR to you showing how to add BVLit in OCaml, together with an example (which produces an SMTLib file which seems reasonable). Do not hesitate to ask questions!

mpardalos committed:
Merge pull request #1 from smtcoq/bitvectors

Adding BVLit in OCaml, with an example

mpardalos committed:
Implement reification for bitvector terms

mpardalos commented:
Implemented reification. I will take a look at compiling hopefully later this week.
